### ISSUE TITLE
[ZG-024] Add bounded multi-stage inverse search with Pareto results

### ISSUE THREAD

**Issue by techrote** [techrote/zaaggenz#25]
# ZG-024 — Bounded multi-stage inverse search

**Milestone:** M2 · **Prerequisites:** ZG-003, ZG-004, ZG-005, ZG-014, ZG-018, ZG-019, ZG-020 · **Lock:** inverse integration

## Goal
Fit useful controllable models/transformations without claiming a unique original production chain or overfitting one mastered excerpt.

## Work
1. Retain the recovered inverse baseline and add staged searches for timing/envelope/root, spectral structure and local texture under explicit parameter bounds.
2. Support waveform/feature/sonority targets with component losses, multi-window support, holdouts and identifiability notes; return Pareto alternatives rather than one opaque optimum.
3. Checkpoint budgets/seeds, support priority cancellation and share deterministic feature/render caches.
4. Evaluate any differentiable DSP adapter only as an optional ablation after the deterministic path is useful.

## Acceptance
- Synthetic known-parameter cases quantify recovery ambiguity and held-out windows reveal overfit.
- Search cannot succeed by silence, lost transients or destructive output clipping unless explicitly allowed/labeled.
- Cancelled/resumed jobs preserve recorded search lineage or declare divergence.
- Results expose trade-offs and can be applied as editable offline recipes without a trained model at playback time.

**Comment by techrote** [comment]
Starting the owner-requested **ZG-024a / Startoff 1** foundation pass on `work/zg024a-inverse-foundations`, based on `6e0153d824608d6a9899c5ea8e8c0b9a29688c98`. Claiming the inverse-integration scope for this pass; shared prerequisite implementations and recovered baseline remain authoritative and unchanged. Scope: baseline provenance, small versioned contracts, deterministic synthetic fixtures, sealed fit/holdout evaluation, decomposed objectives and anti-degeneracy diagnostics, deterministic bounded enumeration, checkpoint/lineage foundations, tests/evidence and `docs/inverse/ZG024_RESEARCH_HANDOFF_1.md`. No advanced optimiser, product defaults or `locked_bloom` changes. This is explicitly a partial serial implementation of ZG-024: #25 stays open after the focused PR merges.

**Comment by techrote** [comment]
**ZG-024a Startoff 1 execution update / concurrent-work preservation:** the original `work/zg024a-inverse-foundations` branch moved from `28bb923...` to `9f85c3ec...` while this pass was being tested locally; the new commit contains a different inverse implementation. I have not overwritten, reset, or deleted that work. This locally verified pass is now isolated on **`work/zg024a-lab-20260914`**, from the same accepted main `6e0153d...` plus the original source-snapshot commit. Local evidence: 44 foundation tests, 373 inherited tests, all 13 recovered-app smoke scripts and browser transport regression passed; six synthetic cases independently rerun deterministically. CI publication is underway. Please avoid parallel publication/merging of another ZG-024a substrate: the final reviewed PR will establish the accepted contract. No main/default changes yet; #25 remains open for subsequent research passes.

**Comment by techrote** [comment]
## ZG-024a / Startoff 1 merged — #25 remains open

Completed the requested foundational pass in **PR #84**, squash-merged as **`7316a6ceeeeac11a48264c925b639d6dfad07f53`**. Verified GitHub reports PR merged, main points to this commit, and this issue is still open. Releasing this pass's inverse-integration scope; no final production optimizer selected.

### Next-pass entry point

Read **[`docs/inverse/ZG024_RESEARCH_HANDOFF_1.md`](https://github.com/techrote/zaaggenz/blob/7316a6ceeeeac11a48264c925b639d6dfad07f53/docs/inverse/ZG024_RESEARCH_HANDOFF_1.md)**, then the laboratory README and recovered-baseline characterization beside it. RAG_INDEX now includes these documents. Search strategies should reuse `FitEvaluator` and the existing contracts/renderer/objectives/scheduler, not receive fixture truth or holdout samples.

### Delivered

Versioned composable contracts; canonical source/target/candidate/render/feature provenance; explicit bounds, seeds and budgets; separate fit/holdout/whole evaluation; decomposed acoustic and sonority losses; independent anti-cheating gates with gain provenance; repeated deterministic rendering and shared bounded caches; research-lane cancellation and verified-prefix checkpoint replay; six synthetic fixture families; transparent bounded cyclic-grid calibration with every candidate retained and Pareto-compatible vectors; read-only preservation of the recovered inverse benchmark; tests, frozen evidence and cross-platform CI.

### Verified acceptance

**All four checks on exact PR head `ff6a5400df802b21e18395a0657d1a17f9ddccff` were green before merge.**

- Foundations Windows + Ubuntu: https://github.com/techrote/zaaggenz/actions/runs/34888914996
- Independent contracts/legacy-equivalence Windows + Ubuntu: https://github.com/techrote/zaaggenz/actions/runs/34888914894

The new suite contains **77 passing tests**; applicable inherited project/jobs/QC/analysis/components/descriptors/DSP/spectral/tuning and browser transport regressions also pass. No calibration, tolerances or tests were weakened for CI.

Downloaded both CI evidence archives and independently verified exact evidence digests and zero differences under the predeclared portable comparison policy. Exact environment-specific report SHA-256 values:

- Windows: `478ac05c2e7ef69eef9fbb52e0de6b342d3a32b6681d103ef81bb775de59795c`
- Ubuntu: `d664d91b56287707b802006378b9cf35313f714901c8315b456124c7844b3662`
- Local independently repeated evidence: `52376b611f758f6bf4773109f2680ff624e796f697dc68099fd34a25680eb035`

### Calibration findings

36 retained logical candidates, 30 fit-eligible, 72 cold verification renders and 36 independent audit renders. Three distinct drive/trim recipes have identical PCM and remain distinct alternatives. Feature/level loss can be zero while waveform relative error is 2. The deliberate late-gain overfit gives fit score zero for all three candidates but holdout scores **1.3065778324 / 0.7065475576 / 0**. All platforms reproduce those outcomes. The unchanged recovered smoke gives 0.3585931590 → 0.3501547175 in 133 evaluations; it is a different objective/target/budget, not a direct optimizer ranking.

Compact permanent calibration: `examples/zg024a_inverse_calibration.json.gz`. Regenerate complete records with:

```text
python tools/inverse_foundation_report.py --out inverse-evidence.json --full-out inverse-full.json --telemetry-out inverse-timing.json --check examples/zg024a_inverse_calibration.json.gz
```

### Explicit limits for research

Initial recipe adapter is single-beat mono legacy-exact. Legacy internal normalization is disclosed; unobservable internal gain factors are null, not invented. Six wrong-root trials activate the derivative-anchor transient gate, exposing phase/periodic-structure sensitivity that needs ablation before broader musical use. Published holdouts must not be treated as indefinitely untouched confirmatory data. Same-environment exact reproducibility is distinct from cross-platform tolerant agreement, and environment changes reject checkpoint continuation.

No changes to `locked_bloom`, product defaults, recovered source or existing contracts/tests. **ZG-024 continues with later research passes; this issue is intentionally not completed or closed.**

**Comment by techrote** [comment]
**Reconciliation:** verified concurrent PR #84 is merged at `7316a6ce...` and is now authoritative for the Startoff 1 contracts/laboratory. Closed my conflicting alternative PR #85 as superseded (not merged), preserved its branch/evidence, and removed its completed temporary publisher. No accepted code was overwritten. A narrowly scoped follow-up on `work/zg024a-association-stability` will carry only the independently reproduced native association-tie/sonority portability finding, tests and handoff addendum into the accepted architecture; no second inverse framework or optimizer research. #25 remains open.

**Comment by techrote** [comment]
Starting **ZG-024b / research pass 2** on `work/zg024b-strategy-research` from merged foundation `7316a6ceeeeac11a48264c925b639d6dfad07f53`.

This pass will keep `zaaggenz_inverse/**` untouched so the frozen ZG-024a calibration remains reproducible. Research code will layer on the fit-only evaluator and compare several deterministic bounded strategies under equal logical budgets, using newly frozen off-grid fixtures plus the existing leakage/non-identifiability sentinels. Candidate generation will not receive fixture truth or holdout samples; audits remain post-selection only.

Planned comparisons: existing cyclic-grid baseline, deterministic pseudo-random coverage, shifted Halton low-discrepancy coverage, and a transparent coordinate/coarse-refinement search. Metrics will retain full component/Pareto records, fit eligibility, holdout diagnostics after selection, physical render/cache cost, and strategy lineage. The goal is evidence about search behavior and failure modes, not selection of a final production optimizer.

I’ll add tests, reproducible evidence, a second research handoff, and cross-platform CI, then merge the focused PR only if applicable checks are green. **#25 remains open.**

**Comment by techrote** [comment]
## ZG-024b / research pass 2 merged — issue remains open

Merged **PR #86** as `cf389e4a54fd925ec1071ae2128c367731e738aa`. GitHub unexpectedly auto-closed #25 at merge despite the PR explicitly containing no closing directive; I immediately reopened it because ZG-024 is still a serial research programme and this pass did not select or implement a final production optimizer.

### What this pass established

The frozen ZG-024a package remains untouched. Research code under `research/zg024b` compares four deterministic bounded strategies through the same fit-only evaluator and validation path: cyclic grid prefix, deterministic SplitMix64 uniform coverage, shifted Halton low-discrepancy coverage, and transparent coordinate refinement with labeled fallback. Each paired comparison uses the same fixture/search seed/logical budget; search receives neither fixture truth nor holdout samples. Verified-prefix replay and explicit strategy lineage are retained.

The preregistered matrix contains 48 core runs (four off-grid/non-identifiable fixtures × three seeds × four methods) plus four independent late-gain sentinel runs. Portable Ubuntu/Windows outcomes match with zero comparison failures under the frozen `1e-7` absolute / `1e-5` relative numerical policy. Exact evidence identities remain environment-specific: Ubuntu `fcac4d9a693ff2185443c6dd8623b126e9cd0d94808dbb8d74734c9f23605a31`, Windows `8ae973ec58caef79b9631da33c3ba362fa715d8c768407287b4e87d71279388a`.

### Main result: no universal optimizer winner

- **Uniform SplitMix** wins all three off-grid envelope seeds and has the best median identifiable parameter error (~0.1538).
- **Shifted Halton** wins all three 3-axis timbre seeds and is the only method with an eligible selected candidate in all 12 core runs.
- **Coordinate refinement** has the best aggregate median selected fit (~0.002284) and holdout (~0.002107), and is strongest on two of three non-identifiable drive/trim seeds plus one nonlinear seed.
- **Grid prefix** wins two nonlinear seeds but has only 58.3% selected-result availability because hard validation rejects many root/timbre proposals.

A near-zero acoustic fit still does not imply generating-parameter recovery: on the nonlinear fixture the grid reaches ~0.000693 fit with normalized raw parameter error ~0.280.

### Gate interaction is now a concrete research blocker

All grid candidates on the off-grid envelope runs are rejected from eligibility, predominantly by `transient_loss`, and several timbre runs are similarly affected. The rejected candidates retain finite decomposed fit measurements. This extends Handoff 1's finding that the current derivative-anchor transient diagnostic can respond to root/phase/periodic-structure changes. The gate was **not weakened** to improve strategy scores. A dedicated transient-gate validation/ablation is now the highest-value next step before treating optimizer eligibility rates as musical evidence.

### Holdout sentinel remains uncompromised

Every method selects a candidate with fitting score **0** because all fit windows occur before the late gain step. Independent selected holdout scores are about **1.30658 / 1.27963 / 1.28689 / 1.30658** for grid / uniform / Halton / coordinate, and every selected candidate is flagged overfit. All nine evaluated candidates remain fit-Pareto-equivalent; holdout evidence never feeds back into search/ranking.

### Final verification

Exact head `44db98add56a00a317554064b3c672de79cd21bd` passed:

- ZG-024b deterministic strategy research Ubuntu + Windows: run `34895865179` — 18 research tests, all 77 frozen ZG-024a inverse tests, complete evidence regeneration against frozen calibration, zero comparison failures.
- Original ZG-024a inverse laboratory Ubuntu + Windows: run `34895865186` — success.
- ZG-002 contracts/legacy equivalence Ubuntu + Windows: run `34895865168` — success.

The one immediately preceding finalization run failed only on a case-sensitive test assertion against the literal calibration note (`CI` vs `ci`); the final commit changes that assertion to compare lowercase. No strategy, fixture, seed, budget, objective, threshold, tolerance, validation rule or frozen outcome was changed after results were observed.

### Next-pass entry point

Read `docs/inverse/ZG024_RESEARCH_HANDOFF_2.md`. It recommends two evidence-first steps: validate/ablate the transient gate using controlled destroyed-attack and root/phase fixtures, then compare explicit staged deterministic search (global low-discrepancy coverage → bounded local refinement) under the same equal-budget/holdout discipline. Larger population, surrogate/Bayesian or differentiable-DSP approaches remain later ablations.

No `locked_bloom`, product defaults, recovered source or `zaaggenz_inverse/**` foundation changes. **ZG-024 / #25 is intentionally open for the next serial pass.**

**Comment by techrote** [comment]
## Newly split ZG-024 follow-up work

Owner requested explicit issues for the currently identified required/recommended fixes rather than leaving them implicit in handoff prose.

- #87 — **required first:** validate/repair `transient_loss` eligibility gating with preregistered phase/root/timbre and genuinely destroyed-attack controls. Do not tune the gate to improve optimizer rankings.
- #90 — repository reconciliation: audit/preserve/clean the ZG-024 branch/PR graveyard without losing unique evidence or rewriting history. May proceed independently where it does not contend for active inverse-integration work.
- #92 — staged deterministic inverse search with fresh confirmatory fixtures. This should follow #87 for any production-selection claim; otherwise the gate must remain report-only and no optimizer should be promoted.

These are child work items of stable ID **ZG-024**, not new programme IDs. PR #84 remains the authoritative ZG-024a foundation and PR #86 the accepted ZG-024b strategy pass. **#25 remains open until the full parent acceptance contract is actually satisfied.**

**Comment by techrote** [comment]
Residual defect audit added #134 under ZG-024: persisted `Candidate` records authenticate render/feature identities but only syntax-check several critical provenance SHAs (`source_revision_id`, `base_recipe_sha256`, `target_asset_sha256`, implementation/environment identity). Those fields can be substituted while candidate/search IDs and eligibility remain unchanged. Repair the persistence evidence envelope before staged work in #92 or inspector integration in #94 treats candidate provenance as authoritative.

**Comment by techrote** [comment]
## ZG-024d staged-search follow-up merged — parent #25 remains open

Child issue #92 is complete via PR #195, merged to `main` as `70d0744477f04485ecedc60d3792f80bd2478f0b` after the exact reviewed head `a5e3c814a1bafdecbeb055d284e4e1a375e7103f` passed ZG-024d, ZG-024b, ZG-024a, ZG-002 and ZG-000 CI.

The preregistered three-stage deterministic strategy research produced a **mixed/no-go** result. No staged optimizer is promoted into production. Development, sealed confirmation and holdout support all failed the frozen production bar, while transient/silence/clipping safety, promotion integrity, exact logical-budget accounting, replay/cancellation, cache identity and cross-platform portability passed.

The remaining ZG-024 blocker is now narrower and evidence-based: achieve reliable **eligible candidate availability plus held-out improvement** on fresh structural/spectral and mixed-texture cases without holdout feedback or weakened safety gates. The next experiment should isolate whether the failure is primarily stage-budget allocation, eligible-parent bottlenecks, or cross-family coupling/identifiability before considering a more opaque optimizer family.

#25 therefore remains open; no parent acceptance claim is made.

**Comment by techrote** [comment]
ZG-024e is now instantiated as child work item #227: **[ZG-024 follow-up] Diagnose eligible-candidate starvation and held-out generalisation failure**. It explicitly owns the inverse-integration shared lock while active and targets the current semantic blocker `research:ZG-024-eligible-candidate-heldout-generalisation`. No parent completion/dependency claim is implied.

**Comment by techrote** [comment]
## ZG-024e completion result

Child #227 / PR #228 has completed the preregistered eligible-starvation/generalisation diagnostic.

Development strongly implicated the strict A-only promotion boundary, but untouched confirmation was **mixed/no-go**: coupled A+B search recovered the targeted structural/spectral availability on 2/3 seeds, yet overall availability was 6/9 for both the coupled intervention and matched factorized null, and all three comparable held-out families exceeded the frozen 1.10 non-regression limit against the best equal-budget flat control.

Safety gates, ineligible-promotion exclusion, exact logical budgets, deterministic/replay/cancellation/cache regressions and Ubuntu/Windows portable evidence remain intact. No production optimiser/default is promoted.

Accordingly #25 remains open and dependency-unsatisfied. Its evidence-backed blocker is narrowed to:

`research:ZG-024-coupled-search-heldout-generalisation`

Future work must use fresh preregistered confirmation; the disclosed ZG-024e `confirm2-*` holdouts are now consumed evidence and must not be tuned against or reused as untouched confirmation.

**Comment by techrote** [comment]
ZG-024e / child #227 is now merged via PR #228 as `5572f7a5fb7dd14b2d9823211cb956221170f982`.

Parent ZG-024 remains **open, research-active, dependency-unsatisfied**. The accepted machine blocker on `main` is `research:ZG-024-coupled-search-heldout-generalisation`. No production inverse optimizer was promoted.
